openedx/core/lib/api/permissions.py

```diff
@@ -21,6 +21,21 @@ def has_permission(self, request, view):
         )
 
 
+class ApiKeyHeaderPermissionIsAuthenticated(ApiKeyHeaderPermission, permissions.IsAuthenticated):
+    """
+    Allow someone to access the view if they have the API key OR they are authenticated.
+
+    See ApiKeyHeaderPermission for more information how the API key portion is implemented.
+    """
+    def has_permission(self, request, view):
+        api_permissions = ApiKeyHeaderPermission.has_permission(self, request, view)
+        print "has api permissions: %d" % api_permissions
+        is_authenticated_permissions = permissions.IsAuthenticated.has_permission(self, request, view)
```

Comment:
What do you think of optimizing this so that you only check and return the less common permission set (presumably `api_permissions`) if the more common one (presumably `is_authenticated_permissions`) is `False`?